### PR TITLE
Fixes expired delegate lookup AB#9666

### PR DESCRIPTION
**Pull request by tiagograf-AOT**
# Fixes or Implements [AB#9666](https://qslvic.visualstudio.com/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/9666)

* [ ] Enhancement
* [x] Bug
* [ ] Documentation

## Description

Fixes the laboratory lookup when a resource delegate is expired.

If you are reviewing this PR, please adhere to the guidelines as [documented](https://github.com/bcgov/healthgateway/wiki/prguidance).

## Testing

* [x] Unit Tests Updated
* [x] Functional Tests Updated
* [ ] Not Required

### Steps to Reproduce

If this is a bug, please provide details on how to reproduce the code.

### UI Changes

NO

### Browsers Tested

* [ ] Chrome
* [ ] Safari
* [ ] Edge
* [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments

Any additional notes or comments that may be relevant.  Include any specialized deployment steps here.

### CODE REVIEW

File: Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs [hunk 1]
```diff
@@ -219,5 +240,27 @@ private bool ValidateObservationDelegate(AuthorizationHandlerContext context, st
...
+        {
+            if (!this.maxDependentAge.HasValue)
+            {
+                this.logger.LogInformation($"Delegate expired check on resource {resourceHDID} skipped as max dependent age is null");
+                return false;
+            }
+
+            if (this.patientService == null)
+            {
+                this.logger.LogError($"Delegate expired check on resource {resourceHDID} failed as patientService is null");
```
Comment (FuriousLlama) [inline Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: I think this should throw since it should never reach this place if going though the correct flow

File: Apps/Common/src/Models/RequestResult.cs [hunk 2]
```diff
@@ -27,6 +27,26 @@ namespace HealthGateway.Common.Models
...
+        public RequestResult()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RequestResult{T}"/> class.
+        /// </summary>
+        /// <param name="resourcePayload">The resource payload.</param>
+        /// <param name="resultStatus">The result status.</param>
+        public RequestResult(
```
Comment (FuriousLlama) [inline Apps/Common/src/Models/RequestResult.cs:42]: Why is this necessary? I dont see a new RequestResult(params)

File: Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs (hunk 1 above)
Comment (sslaws) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: It shouldn't throw but you shouldn't have to do the check... Why are the delegates nullable?  Probably an easy fix when we first introduced the nullable checks but we should change them to non-nullable now and remove the check.

File: Apps/Laboratory/src/appsettings.hgpoc.json [hunk 3]
```diff
@@ -10,6 +10,15 @@
     "Laboratory": {
         "EndPoint": "https://phsahealthgatewayapi-dev.azurewebsites.net/api/v1/LabOrders"
     },
+    "PatientService": {
```
Comment (sslaws) [inline Apps/Laboratory/src/appsettings.hgpoc.json:13]: Nice of you to maintain this file!

**Review by sslaws — CHANGES_REQUESTED** [review]

File: Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs (hunk 1 above)
Comment (FuriousLlama) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: +1

Comment (tiagograf-AOT) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: If this is not nullable, any project that uses the resource authorization handler would have to initialize patient services in their startup which i think its all projects at this point, but only the lab will actually reach this code that checks for delegation. I think theres no harm so far making it not nullable now since the projects won't reach this code

File: Apps/Common/src/Models/RequestResult.cs (hunk 2 above)
Comment (tiagograf-AOT) [reply Apps/Common/src/Models/RequestResult.cs:42]: its not necessary, its a nice to have

File: Apps/Laboratory/src/appsettings.hgpoc.json (hunk 3 above)
Comment (tiagograf-AOT) [reply Apps/Laboratory/src/appsettings.hgpoc.json:13]: if its there we should maintain it, otherwise lets delete it

File: Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs (hunk 1 above)
Comment (tiagograf-AOT) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: i just think its better to get this error instead of a nullreference exception

Comment (tiagograf-AOT) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: I just tested and i was wrong, all projects would have to initialize patient services if I make it not nullable. should i got ahead and enabled it everywhere?

Comment (tiagograf-AOT) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: The error:

Comment (tiagograf-AOT) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Microsoft.AspNetCore.Authorization.IAuthorizationService Lifetime: Transient ImplementationType: Microsoft.AspNetCore.Authorization.DefaultAuthorizationService': Unable to resolve service for type 'HealthGateway.Common.Services.IPatientService' while attempting to activate

Comment (FuriousLlama) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: So the handler is used by all projects, but only some actually reach this code? If that is the case I think throwing an exception will be better than just logging an error since that would imply the startup was not properly set.
Is it possible to have a custom handler for the project? it can inherit from the base one

File: Apps/Common/src/Models/RequestResult.cs (hunk 2 above)
Comment (tiagograf-AOT) [reply Apps/Common/src/Models/RequestResult.cs:42]: its being used in the tests!

Comment (FuriousLlama) [reply Apps/Common/src/Models/RequestResult.cs:42]: Not sure am following. Is it being used?

File: Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs (hunk 1 above)
Comment (tiagograf-AOT) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs:254]: Yes only Observation/Lab will check for delegation so far, I think we plan to do this for other projects in the future, I can look into making a new custom handler

Comment (tiagograf-AOT) [reply Apps/Common/src/AccessManagement/Authorization/Handlers/FhirResourceAuthorizationHandler.cs]: New handler implemented.

File: Apps/Common/src/Models/RequestResult.cs (hunk 2 above)
Comment (tiagograf-AOT) [reply Apps/Common/src/Models/RequestResult.cs:42]: it just makes initializing a mock RequestResult easier (less lines) not a big deal, but i also don't see why this would be bad

**Review by sslaws — APPROVED** [review]

**Review by FuriousLlama — APPROVED** [review]
